lib/remi/data_subjects/gsheet_file.rb

```diff
@@ -0,0 +1,163 @@
+
+    # @return [Array<Extractor::FileSystemEntry>] (Memoized) list of objects in the bucket/prefix
+    def all_entries
+      @all_entries ||= all_entries!
+    end
+
+    # @return [Array<Extractor::FileSystemEntry>] (Memoized) list of objects in the bucket/prefix
+    def all_entries!
+
+      gsheet_entries = get_file_list(@default_folder_id)
+      gsheet_entries.map do |entry|
+        entry = entry.to_h
+        FileSystemEntry.new(
+          pathname:       File.join(@default_folder_id, entry[:name]),
+          create_time:    entry[:created_time],
+          modified_time:  entry[:created_time]
+        )
+      end
+    end
+
+    private def init_gsheet_extractor(*args, credentials:, **kargs)
+
+      @default_folder_id   = credentials.fetch(:folder_id)
+      @oob_uri             = 'urn:ietf:wg:oauth:2.0:oob'
+      @application_name    = 'ITK-ETL'
+
+      @client_secrets_path = File.join(
+        Dir.home,
+        '.credentials/client_secret.json'
+      )
+      @credentials_path = File.join(
+        Dir.home,
+        '.credentials/sheets.googleapis.com-ruby-itk.yaml'
+      )
+      check_secrets_exist(credentials.fetch(:client_id), credentials.fetch(:access_token), credentials.fetch(:refresh_token), credentials.fetch(:client_secret))
+    end
+
+    private def check_secrets_exist(client_id, access_token, refresh_token, client_secret)
+      if client_id.nil? || access_token.nil? || refresh_token.nil? || client_secret.nil?
+        raise "Missing key error"
+      end
+      json_data = <<-JSON
+{"installed":{"client_id":"#{client_id}","project_id":"itk-etl","auth_uri":"https://accounts.google.com/o/oauth2/auth","token_uri":"https://accounts.google.com/o/oauth2/token","auth_provider_x509_cert_url":"https://www.googleapis.com/oauth2/v1/certs","client_secret":"#{client_secret}","redirect_uris":["urn:ietf:wg:oauth:2.0:oob","http://localhost"]}}
+      JSON
+      yaml_file = <<-YAML
+---
+default: '{"client_id":"#{client_id}","access_token":"#{access_token}","refresh_token":"#{refresh_token}","scope":["https://www.googleapis.com/auth/drive","https://www.googleapis.com/auth/spreadsheets"],"expiration_time_millis":1476732343000}'
+      YAML
+
+      if !File.file?(@client_secrets_path)
+        File.write(@client_secrets_path, json_data)
+      end
+      if !File.file?(@credentials_path)
+        File.write(@credentials_path, yaml_file)
+      end
+
+    end
+  end
+
+  class Parser::Gsheet < Parser
+
+    def parse(gs_extract)
+
+      google_vals = gs_extract.data
+      return_hash = nil
+      google_vals.each do |google_val|
+
+        if return_hash.nil?
+          return_hash = Hash.new
+          google_val.values[0].each do |header|
+            return_hash[header.to_sym] = []
+          end
+        end
+
+        keys_temp = return_hash.keys
+
+        google_val.values[1..-1].each do |rows|
+          col_num = 0
+
+          rows.each do |value|
+            return_hash[keys_temp[col_num]].insert(-1,value)
```

Comment:
The Ruby idiom is to use the append operator for this kind of thing (https://ruby-doc.org/core-1.9.3/Array.html#method-i-3C-3C):
````
return_hash[keys_temp[col_num]] << value
````